### PR TITLE
Take a std::size_t item count in the bulk dequeue path

### PR DESCRIPTION
**Pull request by Abhishek9639**
Refs #7576

Second half of group 3 from #7576, following #7582.

> **Note on revision.** The first version of this PR treated `work_items_count_` as a counter that could legitimately be seen below zero and added sign guards around it. @hkaiser pointed out that it cannot, and that the definition should be unsigned instead. He is right, and this now changes the type rather than guarding it. Details in the thread.

## Proposed Changes

  - `libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp`: `get_next_threads` takes a `std::size_t` item count, and `work_items_count_` becomes a `std::atomic<std::uint64_t>`.
  - `libs/core/schedulers/include/hpx/schedulers/lockfree_queue_backends.hpp`: `pop_bulk` takes a `std::size_t` item count.
  - `libs/core/schedulers/include/hpx/schedulers/local_workrequesting_scheduler.hpp`: halve the pending length into a `std::size_t` explicitly.

## Any background context you want to provide?

`get_next_threads` and `pop_bulk` both declared `std::int64_t max_items`, but:

* both return `std::size_t`
* the only caller of `get_next_threads` passes a `std::size_t`
* `pop_bulk` forwards the value to `moodycamel::try_dequeue_bulk`, which takes a `std::size_t`

So the `std::int64_t` sat between a `std::size_t` producer and a `std::size_t` consumer, and it is what MSVC reports as C4244 at 32 bit. Taking a `std::size_t` lets the count keep its type the whole way down.

### `work_items_count_`

The remaining conversion was the queue length itself, and the question was whether it can be negative. It cannot. `work_items_` is pushed to in exactly two places, `schedule_thread` and `move_work_items_from`, and both increment the counter **before** the push:

```cpp
void schedule_thread(threads::thread_id_ref_type thrd, bool other_end = false)
{
    ++work_items_count_.data_;
    // ...
    work_items_.push(thrd.detach(), other_end);
}
```

It is only decremented after a pop has already succeeded, and a pop can only succeed for an item whose push, and so whose increment, already happened. Every decrement is therefore matched by an earlier increment. Had the increment come after the push, a stealer could have popped in between and driven it negative, which is the case I had wrongly assumed.

So the declaration is now unsigned:

```cpp
util::cache_line_data<std::atomic<std::uint64_t>> work_items_count_;
```

`get_queue_length` and `get_pending_queue_length` still return `std::int64_t`, since that is what the rest of the scheduler counter interface uses after #7582, so they convert on the way out. `thread_queue_mc::get_queue_length` already has that shape.

### Something I found while looking at this, not fixed here

The line below the one this PR touches in `local_workrequesting_scheduler` looks wrong, and it is not a typing issue, so I left it alone:

```cpp
task_data thrds(d.num_thread_);
thrds.tasks_.reserve(max_num_to_steal);

#ifdef HPX_HAVE_THREAD_STEALING_COUNTS
    // ... loop that push_back()s up to max_num_to_steal threads
#else
    d.queue_->get_next_threads(
        thrds.tasks_.begin(), thrds.tasks_.size(), false, true);
#endif

    if (!thrds.tasks_.empty())
```

`tasks_` is a `hpx::detail::small_vector`, freshly constructed, and `reserve` only changes capacity, so `thrds.tasks_.size()` is `0` at that point. `get_next_threads` therefore returns immediately and `thrds.tasks_.empty()` is always true, so `handle_steal_request` always falls through to `decline_or_forward_steal_request` and no tasks are ever sent. `HPX_WITH_THREAD_STEALING_COUNTS` defaults to `OFF`, so that `#else` branch is what a default build compiles.

Even with a non zero count it would not be right, since `get_next_threads` writes through the iterator with `*it++` and never grows the vector, so `size()` would stay `0` and the writes would land in raw capacity.

It looks like it wants `resize` and then the real size, or a `back_inserter`, but those two do not mean the same thing for the counts, so I would rather you say which. Happy to do it in a separate PR. Also written up on #7576.

### Verification

I cannot build a 32 bit target locally, so the 32 bit job here is the check for the warnings themselves. What I did check by hand:

* `pop_bulk` has exactly one caller and `get_next_threads` has exactly one, both in the files above.
* Every use of `work_items_count_` in `thread_queue.hpp`, of which there are sixteen, either compares against `0`, increments, decrements, or now converts explicitly on the way out.
* I modelled the new expressions in a standalone file and ran it: `available=7 max=3` gives `3`, `max=100` gives `7`, `max=0` gives `0`, the `min_tasks_to_steal_pending_` guard still fires, and `get_pending_queue_length() / 2` gives `3`. The same file compiles clean at `-m32 -Wall -Wextra -Wshorten-64-to-32 -Werror`, which is clang's equivalent of C4244.

## Checklist

Not all points below apply to all pull requests.

- [ ] I have added a new feature and have added tests to go along with it.
- [ ] I have fixed a bug and have added a regression test.
- [ ] I have added a test using random numbers; I have made sure it uses a seed, and that random numbers generated are valid inputs for the tests.

### CODE REVIEW

**Comment by Abhishek9639** [comment]
@hkaiser this closes out the narrowing side of group 3, so #7582 plus this one covers everything C4244 reported in the schedulers.

There is one thing in here that is not a type change and I would like your call on it rather than guessing. `handle_steal_request` passes `thrds.tasks_.size()` to `get_next_threads` immediately after `reserve`, so the count is zero and the default build never satisfies a steal request through that branch. I wrote it up in the PR description and again on #7576. A real count on its own would not fix it either, because `get_next_threads` writes through the iterator and never grows the vector, so it needs either a `resize` first or a `back_inserter`, and those two do not mean the same thing for the counts. Tell me which you want and I will send it as its own PR.

Nothing here depends on that, so this one can go in on its own if it looks right to you.

Next from #7576 I am picking up `init_threads_count_`, `cached_threads_count_` and the `max_idle_loop_count` reads in `threadmanager.cpp`, which are the same habit in the same file and need no decision from you.

File: libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp [hunk 1]
```diff
@@ -957,21 +957,32 @@ namespace hpx::threads::policies {
...
         template <typename Iterator>
-        std::size_t get_next_threads(Iterator it, std::int64_t max_items,
+        std::size_t get_next_threads(Iterator it, std::size_t max_items,
             bool allow_stealing = false, bool steal = false)
         {
-            std::int64_t const work_items_count =
-                (std::min) (work_items_count_.data_.load(
-                                std::memory_order_relaxed),
-                    max_items);
+            // work_items_count_ is signed because it can be seen below zero
```
Comment (hkaiser) [inline libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp]: I don't think work_items_count can ever be negative. I'd rather fix its definition to use `uint64_t`. I even think there is an assert somewhere making sure it doesn't go below zero.

Comment (Abhishek9639) [reply libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp]: You are right on both counts, and I have changed it to `std::uint64_t`.

I went looking for why I believed it could go negative and could not justify it. The invariant holds structurally: `work_items_` is only pushed to in two places, `schedule_thread` and `move_work_items_from`, and both increment the counter before the push rather than after it. The counter is only decremented after a pop has already succeeded, and a pop can only succeed for an item whose push, and therefore whose increment, already happened. So every decrement is matched by an earlier increment and the value never goes below zero. Had the increment come after the push, a stealer could have popped in between and taken it negative, which is the case I was picturing.

So the declaration is now `util::cache_line_data<std::atomic<std::uint64_t>>`, and `get_next_threads` just takes `std::min` against `max_items` with no sign check.

`get_queue_length` and `get_pending_queue_length` still return `std::int64_t`, since that is the type the rest of the scheduler counter interface uses, so they convert on the way out. That is the same shape `thread_queue_mc::get_queue_length` already has:

```cpp
return static_cast<std::int64_t>(work_items_count_.data_.load(
           std::memory_order_relaxed)) + ...
```

The guard I had added in `local_workrequesting_scheduler` is gone too, and the halving is a plain conversion now.

Two things worth saying:

I could not find the assert you remembered. `grep` for `work_items_count` next to `HPX_ASSERT` across `libs/core/schedulers` and `libs/core/thread_pools` comes back empty, and the asserts near it, such as `HPX_ASSERT(remaining - map_delta >= 0)`, are about the thread map rather than this counter. I did not add one either, because an assert on a counter that other cores mutate concurrently would have to load it separately from the decrement, so it could fire on a perfectly valid interleaving. If you would rather have one anyway, say so and I will add it.

`new_tasks_count_` next to it is still `std::atomic<std::int64_t>`. The same reasoning looks like it applies, but you only asked about this one, so I left it.

Comment (hkaiser) [reply libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp]: > I could not find the assert you remembered. `grep` for `work_items_count` next to `HPX_ASSERT` across `libs/core/schedulers` and `libs/core/thread_pools` comes back empty, and the asserts near it, such as `HPX_ASSERT(remaining - map_delta >= 0)`, are about the thread map rather than this counter. I did not add one either, because an assert on a counter that other cores mutate concurrently would have to load it separately from the decrement, so it could fire on a perfectly valid interleaving. If you would rather have one anyway, say so and I will add it.

Ok, then I misremebered things. However please feel free to add the assert(s).

> `new_tasks_count_` next to it is still `std::atomic<std::int64_t>`. The same reasoning looks like it applies, but you only asked about this one, so I left it.

Sure, please fix that while you're looking at this code.

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!
